extra/to_movement_ops.py

```diff
@@ -0,0 +1,107 @@
+import random
+from tqdm import tqdm
+from extra.optimization.helpers import load_worlds
+from tinygrad.codegen.linearizer import Linearizer
+from tinygrad.ops import LazyOp, MovementOps, TernaryOps, BinaryOps, UnaryOps, ReduceOps, BufferOps, MemBuffer, ConstBuffer
+from tinygrad.helpers import dtypes, prod
+from tinygrad.shape.shapetracker import ShapeTracker
+from tinygrad.shape.view import View
+from tinygrad.shape.symbolic import Node, Variable
+inf, nan = float('inf'), float('nan')
+
+def get_buffer_size(shape, strides, offset, mask):
+  real_shape = tuple(y-x for x,y in mask) if mask else shape
+  offset = offset + sum(strides[i] * (real_shape[i]-1) for i in range(len(strides)) if strides[i]<0)
+  real_offset = offset + (sum(x*st for (x,_),st in zip(mask, strides)) if mask else 0)
+  real_real_shape = [s for s,st in zip(real_shape, strides) if st]
+  strides = [abs(st) for st in strides if st and not isinstance(st, Node)]
+  return real_offset + sum((s-1)*st for s, st in zip(real_real_shape,strides)) + 1
+
+def flatten_view(view: View):
+  real_shape = tuple(y-x for x,y in view.mask) if view.mask else view.shape
+  offset = view.offset + sum(view.strides[i] * (real_shape[i]-1) for i in range(len(view.strides)) if view.strides[i]<0)
+  real_offset = offset + (sum(x*st for (x,_),st in zip(view.mask, view.strides)) if view.mask else 0)
+  real_real_shape = [s for s,st in zip(real_shape, view.strides) if st]
+  strides = [abs(st) for st in view.strides if st and not isinstance(st, Node)]
+  def sort_by_strides(shape, strides): return sorted(zip(shape, strides), key=lambda k: (k[1],-k[0]), reverse=True), sorted(range(len(strides)), key=lambda k: (strides[k],-real_real_shape[k]), reverse=True)
+  ordered_shape_strides, _ = sort_by_strides(real_real_shape, strides)
+  ordered_shape_strides = [list(s) for s in ordered_shape_strides]
+  if strides:
+    i = 0
+    while i < len(ordered_shape_strides):
+      if i<len(ordered_shape_strides)-1 and ordered_shape_strides[i][1] == ordered_shape_strides[i+1][0]*ordered_shape_strides[i+1][1]:
+        ordered_shape_strides[i+1][0] = ordered_shape_strides[i][0]*ordered_shape_strides[i+1][0]
+      else: i += 1
+    flat_shape = [shape_stride[0] for shape_stride in ordered_shape_strides]
```

Comment:
This can be done with `flat_shape, flat_strides = list(zip(*ordered_shape_strides))`, but maybe that's less readable.